src/resmom/renew_creds.c

```diff
@@ -1093,153 +1185,357 @@ send_cred_sisters(job *pjob)
+			ptask->ti_qs.ti_task, pid);
+		log_record(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB, LOG_DEBUG, pjob->ji_qs.ji_jobid, log_buffer);
+
+		ret = PBS_KRB5_OK;
+
+		out:
+			if (fd != -1) {
+				fsync(fd);
+				close(fd);
+			}
+
+		if (local_ticket)
+			free_ticket(ticket, CRED_RENEWAL);
+
+		return ret;
+	}
+
+	close(fd);
+
+	if (fd1 >= 0)
+		close(fd1);
+	if (fd2 >= 0)
+		close(fd2);
+
+	if (setsid() == -1)
+		log_record(PBSEVENT_ERROR, PBS_EVENTCLASS_JOB, LOG_ERR, pjob->ji_qs.ji_jobid, "afslog could not setsid");
+
+	wait_afslog();
+
+	if (local_ticket)
+		free_ticket(ticket, CRED_RENEWAL);
+
+	exit(0);
 }
 
 /**
  * @brief
- * 	removepag - if afs pag is set then it is removed from groups
+ * 	signal_afslog - send signal to the specific job task in order to do
+ *	AFS log or terminate the process on job exit.
  *
+ * @param[in] ptask - job task
+ * @param[in] signal - signal to send
+ *
+ * @return 	int
+ * @retval	PBSGSS_OK on success
+ * @retval	!= PBSGSS_OK otherwise
  */
-void
-removepag()
+int
+signal_afslog(const task *ptask, int signal) {
+	char buf[LOG_BUF_SIZE] ;
+	char pid_file[MAXPATHLEN];
+	FILE *fd;
+	struct stat cache_info;
+
+	if (ptask == NULL)
+		return PBS_KRB5_ERR_INTERNAL;
+
+	job *pjob = ptask->ti_job;
+	if (pjob == NULL)
+		return PBS_KRB5_ERR_INTERNAL;
+
+	if (*pjob->ji_qs.ji_fileprefix != '\0')
+		snprintf(pid_file, sizeof(pid_file), "%s%s_afslog_%8.8X.pid", path_jobs, pjob->ji_qs.ji_fileprefix, (unsigned int)ptask->ti_qs.ti_task);
+	else
+		snprintf(pid_file, sizeof(pid_file), "%s%s_afslog_%8.8X.pid", path_jobs, pjob->ji_qs.ji_jobid, (unsigned int)ptask->ti_qs.ti_task);
+
+	fd = fopen(pid_file, "r");
+	if (fd == NULL) {
+		snprintf(buf, sizeof(buf), "Failed to open pidfile: %s", pid_file);
+		log_err(errno, __func__, buf);
+
+		return PBS_KRB5_ERR_CANT_OPEN_FILE;
+	}
+
+	int pid = 0;
+	if (fscanf(fd, "%d", &pid) < 1) {
+			pid = -1;
+	}
+
+	fclose(fd);
+
+	if (pid >= 0) {
+		if (kill(pid, signal) != 0) {
+			snprintf(buf, sizeof(buf), "afslog for task %8.8X could not send signal %d to PID %d.",
+				(unsigned int)ptask->ti_qs.ti_task, signal, pid);
+			log_err(errno, __func__, buf);
+
+			return PBS_KRB5_ERR_KILL_PROCESS;
+		} else {
+			sprintf(log_buffer, "afslog for task %8.8X, signal %d sent to pid %d",
+				(unsigned int)ptask->ti_qs.ti_task, signal, pid);
+			log_record(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB, LOG_DEBUG, pjob->ji_qs.ji_jobid, log_buffer);
+		}
+	} else {
+		snprintf(buf, sizeof(buf), "afslog for task %8.8X failed to get pid from pidfile: %s",
+			(unsigned int)ptask->ti_qs.ti_task, pid_file);
+		log_err(errno, __func__, buf);
+
+		return PBS_KRB5_ERR_KILL_PROCESS;
+	}
+
+	if (signal != SIGHUP && stat(pid_file, &cache_info) == 0) {
+		unlink(pid_file);
+	}
+
+	return PBS_KRB5_OK;
+}
+
+/**
+ * @brief
+ * 	getpag - recognize afs pag in groups and return the pag.
+ *
+ * @return 	int32_t
+ * @retval	pag > 0 on success
+ * @retval	0 otherwise
+ */
+int32_t
+getpag()
```

Comment:
I'm not certain if you have removed all references to getpag outside of this file. If so, it should become static and removed from the corresponding header file.

Reply:
It is also used in becomeuser_args() because initgroups() removes pag here and we need to save it.